package com.roczyno.project_management.exception;

import com.roczyno.project_management.util.ResponseHandler;
import jakarta.validation.ConstraintViolation;
import jakarta.validation.ConstraintViolationException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.context.request.WebRequest;

import java.time.LocalDateTime;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@ControllerAdvice
public class GlobalException {

	private static final Logger logger = LoggerFactory.getLogger(GlobalException.class);

	@ExceptionHandler({
			UserException.class,
			CommentException.class,
			IssueException.class,
			ProjectException.class,
			InvitationException.class
	})
	public ResponseEntity<Object> handleBadRequestExceptions(RuntimeException ex, WebRequest req) {
		logException(ex);
		List<ErrorDetails> errors = Collections.singletonList(
				new ErrorDetails(ex.getMessage(), req.getDescription(false), LocalDateTime.now())
		);
		return ResponseHandler.errorResponse(errors, HttpStatus.BAD_REQUEST);
	}


	@ExceptionHandler(ConstraintViolationException.class)
	public ResponseEntity<Object> handleValidationException(ConstraintViolationException ex) {
		logException(ex);
		String errorMessage = ex.getConstraintViolations().stream()
				.map(ConstraintViolation::getMessage)
				.collect(Collectors.joining("\n"));
		ErrorDetails errorDetails = new ErrorDetails(errorMessage, "Validation Error", LocalDateTime.now());
		return ResponseHandler.errorResponse(errorDetails, HttpStatus.BAD_REQUEST);
	}

	@ExceptionHandler(MethodArgumentNotValidException.class)
	public ResponseEntity<Object> handleValidationException(MethodArgumentNotValidException ex) {
		logException(ex);
		List<ErrorDetails> errors = ex.getBindingResult().getFieldErrors().stream()
				.map(error -> new ErrorDetails(error.getDefaultMessage(), error.getField(), LocalDateTime.now()))
				.collect(Collectors.toList());
		return ResponseHandler.errorResponse(errors, HttpStatus.BAD_REQUEST);
	}

	@ExceptionHandler(AccessDeniedException.class)
	public ResponseEntity<Object> handleAccessDeniedException(AccessDeniedException ex) {
		logException(ex);
		List<String> errors = Collections.singletonList(ex.getMessage());
		return new ResponseEntity<>(getErrorsMap(errors), new HttpHeaders(), HttpStatus.FORBIDDEN);
	}

	@ExceptionHandler(Exception.class)
	public ResponseEntity<Object> handleOtherExceptions(Exception ex, WebRequest req) {
		logException(ex);
		List<ErrorDetails> errors = Collections.singletonList(
				new ErrorDetails(ex.getMessage(), req.getDescription(false), LocalDateTime.now())
		);
		return ResponseHandler.errorResponse(errors, HttpStatus.INTERNAL_SERVER_ERROR);
	}

	private void logException(Exception ex) {
		logger.error("Exception: ", ex);
	}

	private Map<String, List<String>> getErrorsMap(List<String> errors) {
		Map<String, List<String>> errorResponse = new HashMap<>();
		errorResponse.put("errors", errors);
		return errorResponse;
	}
}
